lib/Sema/SemaExpr.cpp

```diff
@@ -13465,170 +13465,59 @@ bool Sema::CheckBoundsCastBaseType(Expr *E1) {
 ExprResult Sema::ActOnBoundsCastExprBounds(
     Scope *S, SourceLocation OpLoc, tok::TokenKind Kind,
     SourceLocation LAngleBracketLoc, ParsedType D,
-    SourceLocation RAngleBracketLoc, RelativeBoundsClause *RelativeClause,
+    SourceLocation RAngleBracketLoc,
     SourceLocation LParenLoc, SourceLocation RParenLoc, Expr *E1,
-    Expr *bounds) {
+    BoundsExpr *Bounds) {
+  TypeSourceInfo *CastTInfo;
 
-  RangeBoundsExpr *Range = nullptr;
-  TypeSourceInfo *castTInfo;
-
-  QualType DestTy = GetTypeFromParser(D, &castTInfo);
-  ExprResult ParsedBounds = CorrectDelayedTyposInExpr(bounds);
-  SourceLocation TypeLoc = (castTInfo->getTypeLoc()).getBeginLoc();
+  QualType DestTy = GetTypeFromParser(D, &CastTInfo);
+  SourceLocation TypeLoc = (CastTInfo->getTypeLoc()).getBeginLoc();
 
   if (CheckBoundsCastBaseType(E1))
     return ExprError();
 
   if (!DestTy->isCheckedPointerArrayType()) {
-    Diag(TypeLoc, diag::err_bounds_cast_error_with_array_syntax);
+    Diag(TypeLoc, diag::err_bounds_cast_expected_array_ptr);
     return ExprError();
-  } else if ((E1->getType())->isVoidPointerType()) {
-    Diag(E1->getLocStart(),
-         diag::err_typecheck_void_pointer_count_return_bounds);
+  } else if (Bounds->isElementCount() && DestTy->isVoidPointerType()) {
+    Diag(Bounds->getLocStart(),
+         diag::err_typecheck_void_pointer_count_bounds_cast);
     return ExprError();
   }
 
-  if (ParsedBounds.isInvalid())
-    return ExprError();
-
-  if (RelativeClause != nullptr) {
-    RelativeBoundsClause::Kind kind = RelativeClause->getClauseKind();
-    if (((kind == RelativeBoundsClause::Kind::Type) ||
-         (kind == RelativeBoundsClause::Kind::Const)) &&
-        (Range = dyn_cast<RangeBoundsExpr>(ParsedBounds.get()))) {
-      Range->setRelativeBoundsClause(RelativeClause);
-    }
-  }
-
-  return BuildBoundsCastExpr(OpLoc, Kind, castTInfo,
+  return BuildBoundsCastExpr(OpLoc, Kind, CastTInfo,
                              SourceRange(LAngleBracketLoc, RAngleBracketLoc),
-                             SourceRange(LParenLoc, RParenLoc), E1,
-                             dyn_cast<BoundsExpr>(ParsedBounds.get()));
+                             SourceRange(LParenLoc, RParenLoc), E1, Bounds);
 }
 
 ExprResult Sema::ActOnBoundsCastExprSingle(
     Scope *S, SourceLocation OpLoc, tok::TokenKind Kind,
     SourceLocation LAngleBracketLoc, ParsedType D,
-    SourceLocation RAngleBracketLoc, RelativeBoundsClause *RelativeClause,
+    SourceLocation RAngleBracketLoc,
     SourceLocation LParenLoc, SourceLocation RParenLoc, Expr *E1) {
-  TypeSourceInfo *castTInfo;
-  BoundsExpr *bounds;
+  TypeSourceInfo *CastTInfo;
+  BoundsExpr *Bounds;
 
-  QualType DestTy = GetTypeFromParser(D, &castTInfo);
-  SourceLocation TypeLoc = (castTInfo->getTypeLoc()).getBeginLoc();
+  QualType DestTy = GetTypeFromParser(D, &CastTInfo);
```

Comment:
I looked at the code for `GetTypeFromParser` and I think that either way, it will set `CastTInfo` to `nullptr`, so it's okay that it's declared un-initialized. Maybe there should be an `assert` before it's de-referenced, though?

Reply:
Good catch.  We should check that CastTInfo is non-null and use a different source location if it is.  I'll use the location of the '<' symbol given an expression of the form `Op<T>(...)`, where `Op` is a bounds cast operator.  It'll point the programmer in the right direction, although we could be off by any white space between the '<' and T.